echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy

```diff
@@ -0,0 +1,210 @@
+                                       PipelineInitiator pipelineInitiator,
+                                       CounterService counter,
+                                       @Value('${scheduler.compensationJob.windowMs:30000}') long compensationWindowMs,
+                                       @Value('${scheduler.cron.timezone:America/Los_Angeles}') String timeZoneId,
+                                       DateContext dateContext) {
+    this.scheduler = scheduler
+    this.pipelineCache = pipelineCache
+    this.orcaService = orcaService
+    this.pipelineInitiator = pipelineInitiator
+    this.counter = counter
+    this.dateContext = dateContext ?: DateContext.fromCompensationWindow(timeZoneId, compensationWindowMs)
+  }
+
+  @Override
+  void onApplicationEvent(ContextRefreshedEvent event) {
+    if (subscription == null) {
+      subscription = Observable.interval(POLLING_INTERVAL_SECONDS, TimeUnit.SECONDS, scheduler)
+        .doOnNext { onPipelineCacheAwait(it) }
+        .flatMap { tick -> Observable.just(pipelineCache.pipelines) }
+        .doOnError { onPipelineCacheError(it) }
+        .retry()
+        .subscribe { List<Pipeline> pipelines ->
+          if (pipelines.size() == 0) {
+            return
+          }
+          triggerMissedExecutions(pipelines)
+          subscription.unsubscribe()
+        }
+    }
+  }
+
+  void onPipelineCacheAwait(long tick) {
+    log.info('Waiting for pipeline cache to fill')
+  }
+
+  void onPipelineCacheError(Throwable t) {
+    log.error("Error waiting for pipeline cache: ${t.message}")
+  }
+
+  void triggerMissedExecutions(List<Pipeline> pipelines) {
+    log.info('Looking for missed pipeline executions from cron triggers')
+
+    pipelines = pipelines.findAll { !it.disabled }
+    List<Trigger> triggers = getEnabledCronTriggers(pipelines)
+
+    List<String> ids = getPipelineConfigIds(pipelines, triggers)
+    orcaService.getLatestPipelineExecutions(ids)
+      .subscribe({
+        onOrcaResponse(it, pipelines, triggers)
+      }, { onOrcaError(it) })
+  }
+
+  void onOrcaResponse(Collection<PipelineResponse> response, List<Pipeline> pipelines, List<Trigger> triggers) {
+    triggers.each { trigger ->
+      Pipeline pipeline = pipelines.find { it.triggers.contains(trigger) }
+      List<Date> executions = response.findAll { it.pipelineConfigId == pipeline.id }.collect {
+        // A null value is valid; a pipeline that hasn't started won't get re-triggered.
+        it.startTime != null ? new Date(it.startTime) : null
+      }
+
+      if (executions.size() == 0) {
+        return
+      }
+
+      def lastExecution = executions.first()
+      if (lastExecution == null) {
+        return
+      }
+
+      def expr = new CronExpression(trigger.cronExpression)
+      expr.timeZone = dateContext.timeZone
+
+      if (missedExecution(expr, lastExecution, dateContext.triggerWindowFloor, dateContext.now)) {
+        log.info("Triggering missed execution on pipeline ${pipeline}")
```

Comment:
do we need the whole pipeline logged, or is `application`/`pipelineConfigId` sufficient?